packages/react-relay/classic/legacy/store/__tests__/recycleNodesInto-test.js

```diff
@@ -142,6 +164,13 @@ describe('recycleNodesInto', () => {
       expect(recycleNodesInto(prevData, nextData)).toBe(prevData);
     });
 
+    it('recycles arrays with equal objects without mutating frozen `nextData`', () => {
+      const prevData = [{foo: 1}, {bar: 2}];
+      const nextData = [{foo: 1}, {bar: 2}];
+      Object.freeze(nextData);
+      expect(recycleNodesInto(prevData, nextData)).toBe(prevData);
```

Comment:
This test is definitely wrong, correct? If `nextData` is frozen and we don't expect to mutate it, then we would expect `.toBe(nextData)`

Reply:
Hmm, I'm pretty sure this test is correct. It's testing that even if objects are frozen, we still recycle `prevData` object if it is deeply equal to `nextData`. This is important for preventing identical renders from triggering react diffing downstream.

The behavior of frozen objects differs from mutable ones only in that we can't preserve object equity in unchanged branches of the state tree that *does* have changes.

Given this `prevData`:
```
       A1
  B1        B2
C1  C2    C3  C4
```

we can recycle the entire object of `nextData` looks like this:
```
       A1
  B1        B2
C1  C2    C3  C4
```
...whether or not the anything is frozen.

However, if our `nextData` looks like this:
```
       A1
  B1        B2
C1  C2    C3  C5
```
...we can recycle A1 and B1 from `prevData`, but this requires mutating A1, which crashes the whole app if A1 is frozen. If we were really trying to optimize the frozen object case, it would be possible to create a new A1 to hold the `prevData` B1 and `nextData` B2, but since this only effects the dev environment, that's optimization that is not worth doing.
```